builtin/umask.c

```diff
@@ -7,21 +7,146 @@
 #include <sys/stat.h>
 #include "builtin.h"
 
-/*
- * TODO: Implement symbolic mode
- */
-
 static const char umask_usage[] = "usage: umask [-S] [mode]\n";
 
+enum umask_symbolic_state {
+	UMASK_WHO,
+	UMASK_PERM,
+};
+
+static mode_t umask_current_mask(void) {
+	mode_t mask;
+	const mode_t default_mode = 0022;
+
+	mask = umask(default_mode);
+	umask(mask);
+	return mask;
+}
+
+static int umask_update_mode(mode_t *mode, char action, mode_t *perm_mask, mode_t *who_mask) {
```

Comment:
We generally make functions return a `bool`, true indicating success and false indicating failures, when we don't need more.